runtime/bcutil/ROMClassBuilder.cpp

```diff
@@ -248,6 +251,105 @@ ROMClassBuilder::buildROMClass(ROMClassCreationContext *context)
 	return result;
 }
 
+BuildResult
+ROMClassBuilder::handleAnonClassName(J9CfrClassFile *classfile)
+{
+	J9CfrConstantPoolInfo* constantPool = classfile->constantPool;
+	U_32 cpThisClassUTF8Slot = constantPool[classfile->thisClass].slot1;
+	U_32 originalStringLength = constantPool[cpThisClassUTF8Slot].slot1;
+	const char* originalStringBytes = (const char*)constantPool[cpThisClassUTF8Slot].bytes;
+	U_16 newUtfCPEntry = classfile->constantPoolCount - 1; /* last cpEntry is reserved for anonClassName utf */
+	U_32 i = 0;
+	BOOLEAN stringOrNASReferenceToClassName = FALSE;
+	BOOLEAN newCPEntry = TRUE;
+	UDATA newAnonClassNameLength = originalStringLength + 1 + ROM_ADDRESS_LENGTH + 1;
+	BuildResult result = OK;
+	PORT_ACCESS_FROM_PORT(_portLibrary);
+
+	/* Find if there are any Constant_String or CFR_CONSTANT_NameAndType references to the className.
+	 * If there are none we don't need to make a new cpEntry, we can overwrite the existing
+	 * one since the only reference to it is the classRef
+	 */
+	for (i = 0; i < classfile->constantPoolCount; i++) {
+		if ((CFR_CONSTANT_String == constantPool[i].tag)
+			|| (CFR_CONSTANT_NameAndType == constantPool[i].tag)
+		) {
+			if (cpThisClassUTF8Slot == constantPool[i].slot1) {
+				stringOrNASReferenceToClassName = TRUE;
+			}
+		}
+	}
+
+
+	if (!stringOrNASReferenceToClassName) {
+		/* do not need the new cpEntry so fix up classfile->constantPoolCount */
+		newCPEntry = FALSE;
+		newUtfCPEntry = cpThisClassUTF8Slot;
+		classfile->constantPoolCount -= 1;
+	}
+
+	J9CfrConstantPoolInfo *anonClassName = &classfile->constantPool[newUtfCPEntry];
+	/*
+	 * alloc an array for the new name with the following format:
+	 * [className]/[ROMClassAddress]\0
+	 */
+
+	if (0 == _anonClassNameBufferLength || (newAnonClassNameLength > _anonClassNameBufferLength)) {
+		j9mem_free_memory(_anonClassNameBuffer);
+		_anonClassNameBuffer = (U_8 *)j9mem_allocate_memory(newAnonClassNameLength, J9MEM_CATEGORY_CLASSES);
```

Comment:
How is it the name can be put into a reusable buffer instead of becoming part of the rom class?

If it hasn't been already, I'm thinking this code should be tested by creating a build with a hard coded `stringOrNASReferenceToClassName = TRUE` and running all the test suites.

Reply:
Tobi explained the _anonClassNameBuffer is similar to the _classFileBuffer, both buffers are written to the final rom class in `ROMClassWriter::writeROMClass`.

Also the testing with hard coded `stringOrNASReferenceToClassName = TRUE` was done with no problems found.